File: dsa/lc/reverse_linked_list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reverseList(head *ListNode) *ListNode {    
    //pointer to the second node
    if head == nil || head.Next == nil {
        return head
    }
    var prev *ListNode
    for head != nil {
        next := head.Next
        head.Next = prev
        prev = head
        head = next
    }
    return prev
}

File: dsa/lc/linked_list/reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # new head is going to be the original tail of the LL
        # think about how you start and end...
        next_node = None
        prev_node = None
        while head != None:
            if head.next == None:
                head.next = prev_node
                return head
            next_node = head.next
            head.next = prev_node
            prev_node = head
            head = next_node


        return head